// ngx-scrollbar/esm2022/lib/utils/pointer-events-adapter.mjs
import { Directive, effect, inject, untracked, ElementRef, NgZone } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import { NG_SCROLLBAR } from '../utils/scrollbar-base';
import { SCROLLBAR_CONTROL } from '../scrollbar/scrollbar-adapter';
import * as i0 from "@angular/core";
export class PointerEventsAdapter {
    constructor() {
        // Reference to the NgScrollbar component
        this.cmp = inject(NG_SCROLLBAR);
        // Reference to the Scrollbar control component
        this.control = inject(SCROLLBAR_CONTROL);
        // Reference to the Document element
        this.document = inject(DOCUMENT);
        // Reference to angular zone
        this.zone = inject(NgZone);
        // The native element of the directive
        this.nativeElement = inject((ElementRef)).nativeElement;
        effect((onCleanup) => {
            const disableInteraction = this.cmp.disableInteraction();
            untracked(() => {
                if (!disableInteraction) {
                    this.zone.runOutsideAngular(() => {
                        this._pointerEventsSub = this.pointerEvents.subscribe();
                    });
                }
                onCleanup(() => this._pointerEventsSub?.unsubscribe());
            });
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.11", ngImport: i0, type: PointerEventsAdapter, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.11", type: PointerEventsAdapter, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.11", ngImport: i0, type: PointerEventsAdapter, decorators: [{
            type: Directive
        }], ctorParameters: () => [] });
