// ngx-scrollbar/esm2022/lib/async-detection.mjs
import { Directive, inject, effect, untracked, input, NgZone } from '@angular/core';
import { ContentObserver } from '@angular/cdk/observers';
import { throttleTime } from 'rxjs';
import { NgScrollbarExt } from './ng-scrollbar-ext';
import * as i0 from "@angular/core";
export class AsyncDetection {
    constructor() {
        this.scrollbar = inject(NgScrollbarExt, { self: true });
        this.zone = inject(NgZone);
        this.contentObserver = inject(ContentObserver);
        this.asyncDetection = input();
        this.scrollbar.skipInit = true;
        let sub$;
        effect((onCleanup) => {
            const init = this.scrollbar.viewport.initialized();
            const externalViewport = this.scrollbar.externalViewport();
            const externalContentWrapper = this.scrollbar.externalContentWrapper();
            const externalSpacer = this.scrollbar.externalSpacer();
            const asyncDetection = this.asyncDetection();
            untracked(() => {
                let viewportElement;
                let contentWrapperElement;
                this.zone.runOutsideAngular(() => {
                    // The content observer should not be throttled using the same function we use for ResizeObserver,
                    // It should detect the content change asap to attach the scrollbar
                    sub$ = this.contentObserver.observe(this.scrollbar.nativeElement).pipe(throttleTime(100, null, {
                        leading: true,
                        trailing: true
                    })).subscribe(() => {
                        // Search for external viewport
                        viewportElement = this.scrollbar.nativeElement.querySelector(externalViewport);
                        // Search for external content wrapper
                        contentWrapperElement = this.scrollbar.nativeElement.querySelector(externalContentWrapper);
                        this.zone.run(() => {
                            if (!init && viewportElement && contentWrapperElement) {
                                // If an external spacer selector is provided, search for it
                                let spacerElement;
                                if (externalSpacer) {
                                    spacerElement = this.scrollbar.nativeElement.querySelector(externalSpacer);
                                }
                                this.scrollbar.skipInit = false;
                                this.scrollbar.altViewport.set(viewportElement);
                                this.scrollbar.altContentWrapper.set(contentWrapperElement);
                                this.scrollbar.altSpacer.set(spacerElement);
                            }
                            else if (!viewportElement || !contentWrapperElement) {
                                this.scrollbar.viewport.reset();
                            }
                        });
                        if (asyncDetection !== 'auto') {
                            sub$.unsubscribe();
                        }
                    });
                });
                onCleanup(() => sub$?.unsubscribe());
            });
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.11", ngImport: i0, type: AsyncDetection, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.11", type: AsyncDetection, isStandalone: true, selector: "ng-scrollbar[externalViewport][asyncDetection]", inputs: { asyncDetection: { classPropertyName: "asyncDetection", publicName: "asyncDetection", isSignal: true, isRequired: false, transformFunction: null } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.11", ngImport: i0, type: AsyncDetection, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'ng-scrollbar[externalViewport][asyncDetection]'
                }]
        }], ctorParameters: () => [] });
