// ngx-scrollbar/esm2022/lib/sync-spacer.mjs
import { Directive, inject, signal, effect, untracked, NgZone } from '@angular/core';
import { SharedResizeObserver } from '@angular/cdk/observers/private';
import { map } from 'rxjs';
import { NgScrollbarExt } from './ng-scrollbar-ext';
import { filterResizeEntries } from './ng-scrollbar.model';
import { getThrottledStream } from './utils/common';
import * as i0 from "@angular/core";
export class SyncSpacer {
    constructor() {
        this.sharedResizeObserver = inject(SharedResizeObserver);
        this.scrollbar = inject(NgScrollbarExt, { self: true });
        this.zone = inject(NgZone);
        /**
         * A signal used to sync spacer dimension when content dimension changes
         */
        this.spacerDimension = signal({});
        let sub$;
        effect((onCleanup) => {
            const spacerElement = this.scrollbar.spacerElement();
            const contentWrapperElement = this.scrollbar.contentWrapperElement();
            const throttleDuration = this.scrollbar.sensorThrottleTime();
            const disableSensor = this.scrollbar.disableSensor();
            untracked(() => {
                if (!disableSensor && contentWrapperElement && spacerElement) {
                    // Sync spacer dimension with content wrapper dimensions to allow both scrollbars to be displayed
                    this.zone.runOutsideAngular(() => {
                        sub$ = getThrottledStream(this.sharedResizeObserver.observe(contentWrapperElement), throttleDuration).pipe(map((entries) => filterResizeEntries(entries, contentWrapperElement))).subscribe(() => {
                            this.zone.run(() => {
                                // Use animation frame to avoid "ResizeObserver loop completed with undelivered notifications." error
                                requestAnimationFrame(() => {
                                    this.spacerDimension.set({
                                        width: contentWrapperElement.offsetWidth,
                                        height: contentWrapperElement.offsetHeight
                                    });
                                });
                            });
                        });
                    });
                }
                onCleanup(() => sub$?.unsubscribe());
            });
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.11", ngImport: i0, type: SyncSpacer, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.11", type: SyncSpacer, isStandalone: true, selector: "ng-scrollbar[externalViewport][syncSpacer]", host: { properties: { "style.--spacer-width": "spacerDimension().width", "style.--spacer-height": "spacerDimension().height" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.11", ngImport: i0, type: SyncSpacer, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'ng-scrollbar[externalViewport][syncSpacer]',
                    host: {
                        '[style.--spacer-width]': 'spacerDimension().width',
                        '[style.--spacer-height]': 'spacerDimension().height'
                    }
                }]
        }], ctorParameters: () => [] });
